### PR TITLE
testcloud.py: set default user `core` for rhcos

### PR DESCRIPTION
**Pull request by HuijingHei**

### CODE REVIEW

File: tmt/steps/provision/testcloud.py [hunk 1]
```diff
@@ -469,10 +469,13 @@ def start(self) -> None:
...
         # Prepare ssh key
         # TODO: Maybe... some better way to do this?
-        if "coreos" in self.image.lower():
+        if re.search('coreos|rhcos', self.image.lower()):
             self._instance.coreos = True
             # prepare_ssh_key() writes key directly to COREOS_DATA
             self._instance.ssh_path = []
+        # rhcos does not allow root login
+        if "rhcos" in self.image.lower():
+            self.user = "core"
```
Comment (happz) [inline tmt/steps/provision/testcloud.py:478]: I'd advise against overwriting the value user might have provided via the `--user` option. I'd argue it should be respected, something like this, maybe?

```python
if 'rhcos' in self.image.lower() and not self.opt('user'):
    self.user = 'core'
```

Comment (lukaszachy) [reply tmt/steps/provision/testcloud.py:478]: Looks we can do that for also for fedora-coreos - https://docs.fedoraproject.org/en-US/fedora-coreos/authentication/#_default_user

Comment (HuijingHei) [reply tmt/steps/provision/testcloud.py:478]: Updated, thanks for the review.
IMU, fedora-coreos allows `root` login.

**Review by lukaszachy — CHANGES_REQUESTED** [review]
We need to find other place to change this. Verbose mode prints
```
workdir: /var/tmp/tmt/run-015/plans/default/provision
        how: virtual
        order: 50
        user: root
        image:  ....
```
It will be very confusing if tmt prints 'user: root' but later uses 'user: core'

**Comment by lukaszachy** [comment]
Another problem `self.opt('user')` at that place is None if the 'user' is configured in the plan metadata (the `self.get()` of provision plugin will take care of respecting plan vs commandline preference)

So setting user need to happen sooner, maybe in `def go()` after TestcloudGuestData is initialized

**Comment by HuijingHei** [comment]
Test locally with the PR, the default user show `core` now.
```
$ tmt run -vvv
/var/tmp/tmt/run-010
Found 1 plan.

/plans/basic
summary: Basic tests
    discover
        how: shell
        order: 50
        summary: 1 test selected
            /script-00
    provision
        how: virtual
        order: 50
        user: core
```

File: tmt/steps/provision/testcloud.py [hunk 2]
```diff
@@ -638,6 +638,9 @@ def go(self) -> None:
             for key in TestcloudGuestData.keys()
             })
 
+        # rhcos does not allow root login, set default user core
+        if ("rhcos" in data.image.lower() and not self.opt('user')):
```
Comment (happz) [inline tmt/steps/provision/testcloud.py:642]: Nitpicking: wrapping the whole condition with `(` & `)` is not needed in Python :)

**Comment by happz** [comment]
@HuijingHei could you try it out with the following patch? It should handle the "did user request `user`?" slightly better, although there's still an issue we can't fix today (`--user root`).

```diff
diff --git a/tmt/steps/provision/testcloud.py b/tmt/steps/provision/testcloud.py
index 8399a879..d176143c 100644
--- a/tmt/steps/provision/testcloud.py
+++ b/tmt/steps/provision/testcloud.py
@@ -639,8 +639,9 @@ class ProvisionTestcloud(tmt.steps.provision.ProvisionPlugin):
             })
 
         # rhcos does not allow root login, set default user core
-        if not data.user and not self.opt('user') and "rhcos" in data.image.lower():
+        if 'rhcos' in data.image.lower() and data.user == TestcloudGuestData._default('user'):
             data.user = 'core'
+
         # Once plan schema is enforced this won't be necessary
         # click enforces int for cmdline and schema validation
         # will make sure 'int' gets from plan data.
```

**Comment by HuijingHei** [comment]
Thanks @happz ! Updated. Maybe this is not the best way, but it works!

**Comment by happz** [comment]
> Thanks @happz ! Updated. Maybe this is not the best way, but it works!

Unfortunately, it's the best way we have right now :/ Some day, we might have something like `data.is_modified('user')`...

**Review by lukaszachy — APPROVED** [review]

**Review by happz — APPROVED** [review]

**Comment by lukaszachy** [comment]
Adjusted commit message so it can be used as-is for release notes (https://tmt.readthedocs.io/en/stable/contribute.html#commits) and merged.

Thank you for the patch

**Comment by lukaszachy** [comment]
Good point.

However we don't expect rhcos to be used with 'root' right?  I'm afraid that tests may pass with root but once same code is executed as intended (by core user or with sudo) they stop working.

**Comment by cgwalters** [comment]
From my PoV if there were ever a bug caused by `ssh root@rhcos` we would definitely fix it.  Such bugs seem *really* unlikely to me.   OTOH, I suspect 97.2% of tests would not care whether they were invoked as root directly or via sudo.  So whatever's easier here.  I'm just saying it's an option.

**Comment by HuijingHei** [comment]
Good suggestion.

> a config that appends `PermitRootLogin yes` to /etc/ssh/sshd_config

I will update like this.

**Pull request merged**